test/functional/pdo_sqlsrv/MsCommon.inc

```diff
@@ -52,6 +53,374 @@ function connect($options=array())
+        return $colDef;
+    }
+}
+
+
+/**
+ * @return string default column name when a name is not provided in the columnMeta class
+ */
+function get_default_colname( $dataType )
+{
+    $colName = "c_" . str_replace( ",", "_", str_replace( "(", "_", $dataType ));
+    $colName = rtrim( $colName, ")" );
+    return $colName;
+}
+
+
+/**
+ * Create a table
+ * @param object $conn : PDO connection object
+ * @param string $tbname : name of the table to be created
+ * @param array $columnMeta : array of columnMeta objects, which contain metadata for one column
+ */
+function create_table( $conn, $tbname, $columnMeta )
+{
+    try
+    {
+        $dropSql = "IF  EXISTS (SELECT * FROM sys.objects WHERE object_id = OBJECT_ID(N'" . $tbname . "') AND type in (N'U')) DROP TABLE $tbname";
+        $conn->exec( $dropSql );
+        $colDef = "";
+        foreach ( $columnMeta as $meta )
+        {
+            $colDef = $colDef . $meta->getColDef() . ", ";
+        }
+        $colDef = rtrim( $colDef, ", " );
+        $createSql = "CREATE TABLE $tbname ( $colDef )";
+        $conn->exec( $createSql );
+    }
+    catch ( Exception $e )
+    {
+        var_dump( $e );
+        exit;
+    }   
+}
+
+
+/**
+ * class for encapsulating optional parameters for PDOStatement::bindParam
+ */
+class bindParamOption {
+    public $index;      //1-based index of the parameter
+    public $pdoType;    //PDO_PARAM_ type specified for PDOStatement::bindParam
+    public $length;     //length specified for PDOStatement::bindParam
+    public $options;    //options specified for PDOStatement::bindParam
+        
+    function __construct( $index, $pdoType = null, $length = null, $options = null )
+    {
+        $this->index = $index;
+        $this->pdoType = $pdoType;
+        $this->length = $length;
+        $this->options = $options;
+    }
+    /**
+     * @param object $stmt : PDO Statement object
+     * @param mix $var : variable to bind to the SQL statement parameter
+     */
+    function bindWithOp( $stmt, $var )
+    {
+        $stmt->bindParam( $this->index, $var, constant( $this->pdoType ), $this->length, $this->options );
+    }
+}
+
+
+/**
+ * Create a row into a table
+ * @param object $conn : PDO connection object
+ * @param string $tbname : name of the table to be created
+ * @param array $inputs : an associative array column name and its value
+ * @param bool $r : if the row was successfully inserted
+ * $param string $api : PDO_SQLSRV API used for executing the insert query
+ *                      accepted values: "exec", "query", "prepare", "prepareBindParam"
+ * @param array $paramOption : an array of bindParamOptions. Should only be provided if $api is "perpareWithParamOp", if provided, has to be in the same order as $inputs
+ * @return object PDOStatement object of the insert statement
+ */
+function insert_row( $conn, $tbname, $inputs, &$r = null, $api = "query", $paramOption = array() )
+{
```

Comment:
This insert with params is somewhat error prone (say the order of $inputs is different from $paramOption, etc.) Why can't we add the logic to `bindParamOptions` such that it will simply invoke `bindParam()` when defaults are used?

Reply:
Suggestion: split this `insert_row()` into two methods, one for _exec/query_ (using a boolean flag to distinguish the two, and this returns an error when `$keystore` is NOT "none") and the other for _prepare_, which takes the extra arguments plus checking if `$paramOption` is empty, etc.

Reply:
Can you elaborate on "Why can't we add the logic to bindParamOptions such that it will simply invoke bindParam() when defaults are used"? To prevent ordering problem between $inputs and $paramOptions, I've already added an $index member in bindParamOption. If the user does not put $inputs in order, the insert sql itself will adapt to that ordering; $index in bindParamOption matches with the $index'th "?" placeholder in the insertSql.

Reply:
For your suggestion, doesn't that mean the test will need to be different between CE enabled / not enabled? For example, for CE not enabled, the test wants to call insert_row( "query" ). If the same test is used for CE enabled, then the test will return an error, and for it to work it has to use insert_row( "prepareWithParamOp" ).

Reply:
Since you're creating a class bindParamOptions anyway, and it has one function that invokes bindParam() with extra options, it will be nice to add a simple method in the same class that calls bindParam() with defaults. As a result, you don't need two input arguments in which you need to compare its ordering, etc.

Reply:
About insert_row(), based on what you said, why does the user need to specify if it's exec, query, or prepare, etc.?

Reply:
bindParamOption is only used when user wants to specify parameter options ( such as PDO::PARAM_ type, length and encoding ). If the user doesn't want to provide any of these options he/she can simply invoke PDOStatement;:bindParam. Thus I don't see why we need another function that called bindParam() with defaults. As for the "two input arguments", do you mean $stmt and $var? How else can I call bindParam() without these two inputs?

I made a parameter to contain the api being used to call the insert sql because I wanted more flexibility for this function (in fact I think we should include this parameter for all common functions that executes any query). A test can insert using exec, query, or prepare based on the user's preference and what he/she want to test. And as we both know, CE does not work without binding parameters, thus when CE is on, the function automatically chooses to bind param with prepare and execute. This way, the user can use this function to test exec, query or prepare on the non-CE side and the test won't fail when CE is being tested.

Reply:
I meant the two input arguments to insert_row(), `$inputs` and `$paramOption`, which can be combined, changing the latter class to also take an input value of any type, thus eliminating the unnecessary logic to match each input value to its param options

We can therefore rename class bindParamOption as inputParam to make it more generic, having a method called bindParam() which takes care of invoking stmt->bindParam according to the param options

About splitting insert_row() into two methods, I suggested that mainly because "prepare" (with or without options) is quite different from exec/query. In the case when we want transparency, the user can still use insert_row_query but if AE is enabled, we can invoke insert_row_prepare behind the scene

Let's discuss